PW/Data synchronization article

p-woznikowski opened the pull request.

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+- Segment
+- ZOHO
+
+<!-- Again, clarification: do we want to add only those platforms that send customer 
```
Remove: Batch, Intercom, ZOHO, Klaviyo,MoEngage,
Include:
Braze, Zapier,

p-woznikowski replied:
Edited. I've linked the respective integration articles.

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+I'll update the links once these get clarified
+
+Also: how to know which platform sends data to V% and which receives V% data?
```
I needed to review the documentation to get that info. It's not that straightforward since it depends on the implementation and the context of using the data. If the data consists of customer sync, then it's fine. If the solution allows sending any API POST request (Braze/Bloomreach Engagement), it can also be used for syncing.

p-woznikowski replied:
Ok, I think I get it 🙂

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+The following endpoints can be used to synchronize customer data:
+
+- [List Customers](ref:list-customers)
```
I don't feel it makes sense to list all the methods here. We can share here just one link to the object.

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+> 🚧 Product data in the payload alone
+>
+> Unlike customer data, product data provided in the payload alone do not create a new product in Voucherify. Products need to be created or imported.
```
There is a switch called "override" that can be added to a product or SKU. If this switch is set to true for that product, redemption will upsert the product.

p-woznikowski replied:
Added the following:
> In a request payload, you can add an attribute `override` with value `true` to the product object.
>
> The `override` attribute is used to enable storing the product data in Voucherify. If the product does not exist, it will be created with a `source_id`. If the product does exist, the provided values for the name, price, and metadata will replace those already stored in Voucherify.

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+- Validation
+- Redemption
+
+The product validation is based on the API payload and the data stored in Voucherify's product catalog. However, if a product does not exist in Voucherify, the validation will take into consideration the payload alone.
```
We can expand here a bit.

While it's possible to synchronize the products, it might be exhausting and requiring lot's of API requests to keep the inventory up to date. Instead of updating the products frequently, you can rely on sending the product data in the qualification, validation, redemption and create order requests only. The verification of the eligibility will consider the data you shared in the request without the need to keep it updated.

p-woznikowski replied:
I've rephrased the whole Section. I'm not sure if this is `good enough` or requires some further polishing.

> Products can be stored in Voucherify but it is not required. The product validation takes place during the following actions:
> - Create order
> - Qualification
> - Redemption
> - Validation
> 
> The product validation is based on the API payload and the data stored in Voucherify's product inventory.
> 
> However, if a product does not exist in Voucherify, the validation will use the product data included in the payload. Even if the product does exist in Voucherify, the payload data are used instead of those stored in the system. This method can be used to avoid frequent API calls to keep the product inventory up to date.
> 
> > 🚧 Product data in the request
> >
> > Unlike customer data, product data provided in the request payload alone do not create a new product in Voucherify. If you want to store products in Voucherify, they need to be created manually in the dashboard or imported with a CSV file.
> > 
> > Also, even though Voucherify uses the payload data instead of the data stored in the system, the product data is not upserted through requests. Products can be updated through product edit function or by importing a CSV file.
> >
> > To enable product updates through the payload, use the [override attribute](#overriding-product-data).
> 
> The products in Voucherify can be grouped into static or dynamic collections based on their parameters. Collections can be used in validation rules to model product-specific limits and loyalty-earning rules.

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+The override set to `true` is used to store the product information in the system. If the product does not exist, it will be created with a `source_id`. If the product does exist, the provided values for the name, price, and metadata will replace those already stored in Voucherify.
+
+<!-- I don't get it; how does it work? Why is it in response? Also, in the HS article on products, we read 
```
It's not in the response. It's in the request. If it's used in the redemption or create order request, the product is updated.

p-woznikowski replied:
Clarified further on Slack. Current docs are misleading in this regard 🙁

kasjankotyniarsp commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+<!-- I don't get it; how does it work? Why is it in response? Also, in the HS article on products, we read 
+
+> Override/Update products' names in Voucherify using this method. Data will be updated for each product included in the CSV file whose source_id matches a Source id in Voucherify. No other data can be updated other than the product name.
```
I don't understand the comment :)

p-woznikowski replied:
It's something which I found in HS docs and which increased my confusion 😉

p-woznikowski commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+The customer data does not have to be stored in Voucherify before a validation or redemption request is made. Voucherify can use the customer data from the request to create a new customer with a unique `id` field.
+
+However, the customer data must exist in Voucherify before a distribution is made.
```
Rephrased it as follows:
> However, the customer data must exist in Voucherify before a distribution is made. Also, distribution can be made on the basis of the customer's attributes and sent out to customers from a given segment. If the customer data is not up to date, the distribution messages sent manually can reach the wrong audience.

p-woznikowski commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+Every customer has unique identification data in the `source_id` field. The `source_id` can be a customer ID or email from a CRM system, database, or a third-party service.  It can be imported into Voucherify from your [CRM database]() or a CSV file <!-- link when ready! -->
+
+> 🚧 Customer data synchronization and `source_id`
+>
+> If your data is to be synchronized based on the `source_id` of the customer, you need to import the `source_id` when the customer is uploaded for the first time. You will not be able to change or update `source_id` later on.
+
```
Right...copy-pasted that bit kinda thoughtlessly 🤦‍♂️

p-woznikowski commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+- order creation
+- voucher publication
+- validation <!-- IMO should be here-->
+- redemption
```
Added

p-woznikowski commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+- validation <!-- IMO should be here-->
+- redemption
+
+<!-- Redemption... but this is to be deprecated, right? so should we actually link/mention those
```
This section reads as follows:
> The following actions upsert customer data:
> - [Order creation](ref:create-order)
> - [Stackable discount redemption](ref:redeem-stacked-discounts)
> - [Stackable discount redemption (client side)](ref:redeem-stacked-discounts-client-side)
> - [Track custom event](ref:track-custom-event)
> - [Voucher publication](ref:create-publication)

p-woznikowski commented on docs/guides/implementation_blueprint/data-synchronization.md:
```diff
@@ -0,0 +1,200 @@
+
+-->
+
+<!-- ##### Order creation
```
At first, I thought it would need more in-depth approach, but then I changed my mind. Nevertheless, I've left it there for future use in case of feedback that these needs elaboration. Since you find the section fine, I'm going to remove those 🙂

kasjankotyniarsp commented on docs/guides/development/data-synchronization.md:
```diff
@@ -0,0 +1,169 @@
+The following actions upsert customer data:
+- [Order creation](ref:create-order)
+- [Stackable discount redemption](ref:redeem-stacked-discounts)
+- [Stackable discount redemption (client side)](ref:redeem-stacked-discounts-client-side)
```
We should also add client-side publications if we provide the client side stackable redemption

p-woznikowski replied:
Solved on Slack; no change made

kasjankotyniarsp commented on docs/guides/development/data-synchronization.md:
```diff
@@ -0,0 +1,169 @@
+
+Read our [customer import guide](https://support.voucherify.io/article/67-how-to-import-my-customers#csv-import "How to import my customers?") to learn more.
+
+<!-- Should I write a full step-by-step manual for this? -->
```
Maybe in the future when we will be removing articles from success.voucherify.io

p-woznikowski replied:
Sure thing!

But this particular import guide is on `support.voucherify.io`, so it will be staying for longer ;)

kasjankotyniarsp commented on docs/guides/development/data-synchronization.md:
```diff
@@ -0,0 +1,169 @@
+
+To update customer data, use [Update customer](ref:update-customer) API endpoint. The [Create customer](ref:create-customer) API endpoint can be also used to update customer data.
+
+<!-- Simply repeat the endpoints? -->
```
It's fine as it is.

kasjankotyniarsp commented on docs/guides/development/data-synchronization.md:
```diff
@@ -0,0 +1,169 @@
+
+> 📘 Upserting customer data through API
+>
+> All endpoints that have a customer object can be used to upsert customer data.
```
No, this one is not 100% true. I think we can remove this section, since it's copying lines 64-69

p-woznikowski replied:
Ok, removed

kasjankotyniarsp commented on docs/guides/development/data-synchronization.md:
```diff
@@ -0,0 +1,169 @@
+## Product synchronization
+
+Products can be stored in Voucherify but it is not required. The product validation<!-- validation? --> takes place during the following actions:
+- Create order
```
Creating an order does not validate the product. However, creating an order triggers distributions and earning rules.

p-woznikowski replied:
Resolved on slack

kasjankotyniarsp commented on docs/guides/development/data-synchronization.md:
```diff
@@ -0,0 +1,169 @@
+
+#### Import orders
+
+The [import orders](ref:import-orders) endpoint is used to import only historical orders into Voucherify. For on-going synchronization, the [Create order](ref:create-order) and  [update order](ref:update-order) endpoints should be used.
```
It's worth to mention, that importing orders in bulk won't trigger distributions and earning rules.

p-woznikowski replied:
I've added that sentence right at the end of this paragraph

p-woznikowski commented:
Changes:
- Added the [Data synchronization](https://docs.voucherify.io/v2018-08-01-pw-571/docs/data-synchronization) article
- Changed hyphens to ndashes in the [Introduction](https://docs.voucherify.io/v2018-08-01-pw-571/docs/introduction) article
- Added a link to [Batch](https://support.voucherify.io/article/614-batch-integration) integration article in the [Distributions](https://docs.voucherify.io/v2018-08-01-pw-571/docs/distributions) article
